File: bKap3_3_InSampleRegressionAll.m
% Script for performing the in-sample regressions in GWZ

% Clear console
clear; clc; close all;

% Set paths
sOldPath = path;
sDataPath = './DATA/Liedtke/US/';
sResultsPath = './RESULTS/GWZ/';
addpath('./Utils/');

%% Load data
% Macro panel produced by DATA/Liedtke/aggregate.py.
% CSV layout: col 1 = observation_date, col 2 = target (inflation),
%             col 3..end = predictors (already lag-aligned in Python).
tData     = readtable([sDataPath, 'aggregated.csv']);
cAllNames = tData.Properties.VariableNames;

% Parse dates (first column)
dtDates = tData{:,1};
if ~isdatetime(dtDates)
    dtDates = datetime(dtDates);
end

% Target (col 2) and predictors (col 3..end)
mX       = tData{:,3:end};          % predictors
vY       = tData{:,2};              % inflation target
cXnamesM = cAllNames(3:end)';       % predictor names as an Nx1 cell

%% Data preprocessing
% Predictive lag (periods). The predictors are already reporting-lag aligned
% in Python; this is the additional one-period predictive lag.
iTimeLag = 1;

% Determine dimensions
[iNumObs, iNumAssets]       = size(vY);
[iNumObsP, iNumPredictors]  = size(mX);

% Lag the predictors so that they are known one period ahead of the target
mXlag       = [NaN(iTimeLag, iNumPredictors); mX(1:end-iTimeLag,:)];

%% In-Sample analysis
% Initialize memory
mYhat           = NaN(iNumObs, iNumPredictors);     % Predicted values
dtDateSample    = NaT(iNumPredictors, 2);           % Start and end date of analysis
mBeta           = NaN(iNumPredictors, 2);           % Beta coefficients
mBetaT          = NaN(iNumPredictors, 2);           % t-Statistics of beta coefficients
vR2             = NaN(iNumPredictors, 1);           % R2

% Perform in-sample regression for each predictor
for iIdxP = 1:iNumPredictors
    % Print progress
    fprintf('Estimate in-sample %s\n', cXnamesM{iIdxP})

    % Get data / copy data
    vXtemp      = mXlag(:,iIdxP);
    vYtemp      = vY;
    dtDatesTemp = dtDates;

    % Remove missing values
    lIsNaN = isnan(vXtemp) | isnan(vYtemp);
    vYtemp(lIsNaN)      = [];
    vXtemp(lIsNaN)      = [];
    dtDatesTemp(lIsNaN) = [];

    % Save start and end date
    dtDateSample(iIdxP, :) = [dtDatesTemp(1) , dtDatesTemp(end)];

    % z-Transformation of predictor
    vXtemp = (vXtemp - mean(vXtemp))./std(vXtemp);

    % Regression with intercept
    rResults    = regstats(vYtemp, vXtemp, 'linear', {'tstat','rsquare'});
    vBetaTemp   = rResults.tstat.beta;      % coefficient estimates
    vBetaTempT  = rResults.tstat.t;         % t-Statistic estimates
    dR2         = rResults.rsquare;         % R2

    % Add constant
    mXtemp = [ones(size(vXtemp,1),1), vXtemp];

    % In-sample prediction
    mYhat(~lIsNaN,iIdxP)  = mXtemp * vBetaTemp;

    % Save beta, t-statistics, and R2
    mBeta(iIdxP,:)    = vBetaTemp; 
    mBetaT(iIdxP,:)   = vBetaTempT; 
    vR2(iIdxP)        = dR2;
end

% === Create table
% Merge all results and round to two digits
mResults = round([mBeta(:,2)*100, mBetaT(:,2), vR2 * 100],2);

% Make to cell
cTable3 = [cXnamesM, sprintfc('%.2f', mResults)];

% Add column header
cTable3 = [{'Predictor','Beta', 'Beta T', 'R2'}; cTable3];

% === Save results
sFilename = [sResultsPath,'InSampleResults.mat'];
save(sFilename, "cTable3", 'mBetaT','mBeta','vR2','cXnamesM','mYhat');

% % Compare with actual results
% tResultsGWZ = readtable('./RESULTS/GWZ/ResultsGWZ.xlsx');
% 
% % Match
% [cVars, idxA, idxB] = intersect(tResultsGWZ.Name, cXnamesM);
% scatter(tResultsGWZ.ISR2(idxA),vR2(idxB)*100,'black','filled');
% lsline
% % title('Replikation der in-sample Ergebnisse','FontSize',12)
% ylabel('Replicated $R^2$ (in \%)','FontSize',12,'Interpreter','latex');
% xlabel('GWZ $R^2$ (in \%)','FontSize',12,'Interpreter','latex');
% box off



% Restore path
path(sOldPath);

File: python/bKap3_3_InSampleRegressionAll.py
"""
bKap3_3_InSampleRegressionAll.py
Python port of bKap3_3_InSampleRegressionAll.m.

In-sample univariate predictive regression of inflation on EACH (z-standardised)
predictor separately. Builds the beta / t-stat / R2 table and the in-sample
fitted values, and saves them (Python copy of InSampleResults.mat).

Run from the repo root:  python python/bKap3_3_InSampleRegressionAll.py
"""

import os
import numpy as np
import pandas as pd
import scipy.io as sio

import models

DATA_PATH    = "./DATA/Liedtke/US/aggregated.csv"
RESULTS_PATH = "./RESULTS/GWZ/"

# =========================================================================
#  Load data
# =========================================================================
tData = pd.read_csv(DATA_PATH)
cAllNames = list(tData.columns)
dtDates = pd.to_datetime(tData.iloc[:, 0])
mX = tData.iloc[:, 2:].to_numpy(float)
vY = tData.iloc[:, 1].to_numpy(float)
cXnamesM = cAllNames[2:]

iTimeLag = 1
iNumObs, iNumPredictors = mX.shape

# Lag the predictors so they are known one period ahead of the target
mXlag = np.full_like(mX, np.nan)
mXlag[iTimeLag:, :] = mX[:-iTimeLag, :]

# =========================================================================
#  In-sample analysis, one predictor at a time
# =========================================================================
mYhat = np.full((iNumObs, iNumPredictors), np.nan)
mBeta = np.full((iNumPredictors, 2), np.nan)
mBetaT = np.full((iNumPredictors, 2), np.nan)
vR2 = np.full(iNumPredictors, np.nan)
dtDateSample = [[None, None] for _ in range(iNumPredictors)]

for j in range(iNumPredictors):
    print(f"Estimate in-sample {cXnamesM[j]}")
    vXtemp = mXlag[:, j]
    lIsNaN = np.isnan(vXtemp) | np.isnan(vY)
    keep = ~lIsNaN
    vYk = vY[keep]
    vXk = vXtemp[keep]

    dtDateSample[j] = [dtDates[keep].iloc[0], dtDates[keep].iloc[-1]]

    # z-transformation (sample std, ddof=1)
    vXz = (vXk - vXk.mean()) / vXk.std(ddof=1)

    reg = models.ols_regress(vYk, vXz)
    mBeta[j, :] = reg["beta"]
    mBetaT[j, :] = reg["t"]
    vR2[j] = reg["R2"]
    mYhat[keep, j] = reg["yhat"]

# =========================================================================
#  Build the result table (cTable3) and save
# =========================================================================
mResults = np.round(np.column_stack([mBeta[:, 1] * 100, mBetaT[:, 1], vR2 * 100]), 2)
tbl = pd.DataFrame(mResults, columns=["Beta", "Beta T", "R2"])
tbl.insert(0, "Predictor", cXnamesM)
print("\n========== In-sample results ==========")
print(tbl.to_string(index=False))

os.makedirs(RESULTS_PATH, exist_ok=True)
sio.savemat(os.path.join(RESULTS_PATH, "InSampleResults_py.mat"), {
    "cXnamesM": np.array(cXnamesM, dtype=object).reshape(-1, 1),
    "mBeta": mBeta, "mBetaT": mBetaT, "vR2": vR2.reshape(-1, 1), "mYhat": mYhat,
})
